src/compiler/checker.ts

```diff
@@ -12529,6 +12529,14 @@ namespace ts {
                 if (target.flags & TypeFlags.Simplifiable) {
                     target = getSimplifiedType(target, /*writing*/ true);
                 }
+                if (source.flags & TypeFlags.Substitution) {
+                    source = (<SubstitutionType>source).substitute;
+                }
+                if (target.flags & TypeFlags.Substitution) {
+                    target = (<SubstitutionType>target).typeVariable;
+                }
+                Debug.assert(!(source.flags & TypeFlags.Substitution), "Source type was a substitution - substitutes are unhandled in relationship checking");
+                Debug.assert(!(target.flags & TypeFlags.Substitution), "Target type was a substitution - substitutes are unhandled in relationship checking");

```

Comment:
Perhaps this instead:

```ts
while (true) {
    const s = source.flags & TypeFlags.Substitution ? source = (<SubstitutionType>source).substitute :
        source.flags & TypeFlags.Simplifiable ? getSimplifiedType(source, /*writing*/ false) :
        source;
    const t = target.flags & TypeFlags.Substitution ? (<SubstitutionType>target).typeVariable :
        target.flags & TypeFlags.Simplifiable ? getSimplifiedType(target, /*writing*/ true) :
        target;
    if (s === source && t === target) break;
        source = s;
        target = t;
    }
}
```

Reply:
@ahejlsberg done - let's see if anything stacks out on DT~